Magefile.go

```diff
@@ -0,0 +1,231 @@
+var Default = Build
+
+var Aliases = map[string]interface{}{
+	"test": TestRace,
+}
+
+func init() {
+	if exe := os.Getenv("GOEXE"); exe != "" {
+		goexe = exe
+	}
+
+	// We want to use Go 1.11 modules even if the source lives inside GOPATH.
+	// The default is "auto".
+	os.Setenv("GO111MODULE", "on")
+}
+
+// Fmt run gofmt linter
+func Fmt() error {
+	if !isGoLatest() {
+		return nil
+	}
+	pkgs, err := packages()
+	if err != nil {
+		return err
+	}
+	failed := false
+	first := true
+	for _, pkg := range pkgs {
+		files, err := filepath.Glob(filepath.Join(pkg, "*.go"))
+		if err != nil {
+			return nil
+		}
+		for _, f := range files {
+			// gofmt doesn't exit with non-zero when it finds unformatted code
+			// so we have to explicitly look for output, and if we find any, we
+			// should fail this target.
+			s, err := sh.Output("gofmt", "-l", f)
+			if err != nil {
+				fmt.Printf("ERROR: running gofmt on %q: %v\n", f, err)
+				failed = true
+			}
+			if s != "" {
+				if first {
+					fmt.Println("The following files are not gofmt'ed:")
+					first = false
+				}
+				failed = true
+				fmt.Println(s)
+			}
+		}
+	}
+	if failed {
+		return errors.New("improperly formatted go files")
+	}
+	return nil
+}
+
+//  Vet run go vet linter
+func Vet() error {
+	if err := sh.Run(goexe, "vet", "./..."); err != nil {
+		return fmt.Errorf("error running go vet: %v", err)
+	}
+	return nil
+}
+
+// TestRace run tests with race detector
+func TestRace() error {
+	env := map[string]string{"GOFLAGS": testGoFlags()}
+	return runCmd(env, goexe, "test", "-race", "./...", buildFlags(), "-tags", buildTags())
+}
+
+// TestCoverHTML generates test coverage report
+func TestCoverHTML() error {
+	const (
+		coverAll = "coverage-all.out"
+		cover    = "coverage.out"
+	)
+	f, err := os.Create(coverAll)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	if _, err := f.Write([]byte("mode: count")); err != nil {
+		return err
+	}
+	pkgs, err := packages()
+	if err != nil {
+		return err
+	}
+	for _, pkg := range pkgs {
+		if err := sh.Run(goexe, "test", "-coverprofile="+cover, pkg); err != nil {
+			return err
+		}
+		b, err := ioutil.ReadFile(cover)
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+			return err
+		}
+		idx := bytes.Index(b, []byte{'\n'})
+		b = b[idx+1:]
+		if _, err := f.Write(b); err != nil {
+			return err
+		}
+	}
+	if err := f.Close(); err != nil {
+		return err
+	}
+	return sh.Run(goexe, "tool", "cover", "-html="+coverAll)
+}
+
+// Build run linters, tests, download modules and install
+func Build() error {
+	if strings.Contains(runtime.Version(), "1.8") {
+		// Go 1.8 doesn't play along with go test ./... and /vendor.
+		// We could fix that, but that would take time.
+		fmt.Printf("Skip Build on %s\n", runtime.Version())
+		return nil
+	}
+
+	mg.Deps(Fmt, Vet, TestRace)
+	if err := sh.Run("go", "mod", "download"); err != nil {
+		return err
+	}
+	return sh.Run("go", "install", "./...")
```

Comment:
wondering what it'll build, install and run in our case ? examples ? what one(s) ?
this is a library, only what use it can be installed. We could eventually want to build "examples" but then we would have to make those examples a bit more "unified"

Reply:
@jbguerraz I agree, I'm working on issue #36 where I will introduce a Dockerfile and run the examples as a part of integration tests. I can exclude `install` from this PR. This PR is only aimed at resolving #35.

Reply:
let's keep it simple as @jbguerraz suggested and keep the scope of the PR w.r.t what we have now. i.e perhaps just call `test` and `vet`, i believe i saw some errors when i ran `lint` which we  can tackle in a different PR.